runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -225,49 +376,73 @@ public void processElement(ProcessContext context) {
      */
     private StateTag<Object, ValueState<RestrictionT>> restrictionTag;
 
+    private StateInternalsFactory<String> stateInternalsFactory;
+    private TimerInternalsFactory<String> timerInternalsFactory;
+    private DoFnRunners.OutputManager outputManager;
+    private TupleTag<OutputT> mainOutputTag;
+
     private final DoFn<InputT, OutputT> fn;
     private final Coder<? extends BoundedWindow> windowCoder;
 
     private transient DoFnInvoker<InputT, OutputT> invoker;
 
-    ProcessFn(
+    public ProcessFn(
         DoFn<InputT, OutputT> fn,
         Coder<InputT> elementCoder,
         Coder<RestrictionT> restrictionCoder,
         Coder<? extends BoundedWindow> windowCoder) {
       this.fn = fn;
       this.windowCoder = windowCoder;
-      elementTag =
+      this.elementTag =
           StateTags.value("element", WindowedValue.getFullCoder(elementCoder, this.windowCoder));
-      DoFnInvoker<InputT, OutputT> invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
-      restrictionTag = StateTags.value("restriction", restrictionCoder);
+      this.restrictionTag = StateTags.value("restriction", restrictionCoder);
     }
 
-    @Override
+    public void setStateInternalsFactory(StateInternalsFactory<String> stateInternalsFactory) {
+      this.stateInternalsFactory = stateInternalsFactory;
+    }
+
+    public void setTimerInternalsFactory(TimerInternalsFactory<String> timerInternalsFactory) {
+      this.timerInternalsFactory = timerInternalsFactory;
+    }
+
+    public void setOutputManager(DoFnRunners.OutputManager outputManager) {
```

Comment:
A `DoFn` should not depend on `DoFnRunner` and definitely shouldn't have a handle on `DoFnRunners.OutputManager`. A `DoFnRunner` is just one way of running a `DoFn` that we provide as a convenience.